### DISCUSSION TITLE
How to invoke model to draw contours?

### DISCUSSION THREAD

**Discussion by csmayo157** [microsoft/InnerEye-DeepLearning#564] [Q&A] [answered]
HI All, 
    There are still some pieces that are being put into place on our Azure instance that will give us an IDE for running InnerEye and editing files. 

    In the mean time, we are working on the local system that we have set up with GPUs. We’ve been able to 
1)	Install and run the “Hello World” version of the InnerEye-Deep Learning
2)	Run the InnerEye-Create data set to take our DICOM RT and put them into NIFTI format
3)	Run the InnerEye-Deep Learning on the NIFTI files we imported. 

We’re not finding what we need to do for the next step

New Uncontoured NIFTI file + InnerEyeModel -> NIFTI with contours

Can you give pointers to where we can see how to do that?

**Comment by ant0nsc** [comment]
There’s a section in [docs/building_models.md](https://github.com/microsoft/InnerEye-DeepLearning/blob/main/docs/building_models.md) that describes what you need to do:

Running a registered AzureML model on a single image on the local disk
To submit an AzureML run to apply a model to a single image on your local disc, you can use the script submit_for_inference.py, with a command of this form:
```
python InnerEye/Scripts/submit_for_inference.py --image_file ~/somewhere/ct.nii.gz --model_id Prostate:555  --settings ../somewhere_else/settings.yml --download_folder ~/my_existing_folder
```

**Reply by furtheraway** [reply]
I wonder how to do inference from a locally trained model, like the one produced by HelloWorld. I didn't see how to find the model_id for it. Thank you!

**Reply by ant0nsc** [reply]
Hi @furtheraway, this is not supported at present. Our team operates cloud-first.

**Reply by furtheraway** [reply]
Thank you Anton @ant0nsc ! I tried the command you mentioned, but run into the following error message:
```
(InnerEye2) [yuayao@LNX00033 InnerEye-DeepLearning]$ pwd
/home/yuayao/source/2022/InnerEye-DeepLearning
(InnerEye2) [yuayao@LNX00033 InnerEye-DeepLearning]$ python InnerEye/ML/submit_for_inference.py --image_file /home/yuayao/datasets/nifti/pat0.zip --model_id LungExample:1 --download_folder /home/yuayao/source/2022/download/ --use_dicom true
Failure while loading azureml_run_type_providers. Failed to load entrypoint automl = azureml.train.automl.run:AutoMLRun._from_run_dto with exception (cloudpickle 2.0.0 (/home/yuayao/.conda/envs/InnerEye2/lib/python3.7/site-packages), Requirement.parse('cloudpickle<2.0.0,>=1.1.0'), {'azureml-dataprep'}).
Traceback (most recent call last):
  File "InnerEye/ML/submit_for_inference.py", line 15, in <module>
    from health_azure import create_run_configuration, submit_run
ModuleNotFoundError: No module named 'health_azure'
(InnerEye2) [yuayao@LNX00033 InnerEye-DeepLearning]$
```

I am new in using python and not familiar with how python locates the modules it needs.

However, I run the following runner.py without issue, which also reference the health_azure module.
` python InnerEye/ML/runner.py --model=HelloWorld --azureml`

**Reply by ant0nsc** [reply]
@furtheraway, It appears your conda environment does not yet have the required packages. You can try two things: 
* `pip install hi-ml-azure`
* Alternatively, look at the code at the start of `runner.py`, that adds the paths to the git submodules that contain `health_ml`. Copy those bits into `submit_for_inference`.

**Reply by furtheraway** [reply]
Thank you very much @ant0nsc !

I copied the following section from runner.py into submit_for_inference.py, and it works now.

```
# Workaround for an issue with how AzureML and Pytorch Lightning interact: When spawning additional processes for DDP,
# the working directory is not correctly picked up in sys.path
print(f"Starting InnerEye runner at {sys.argv[0]}")
innereye_root = Path(__file__).absolute().parent.parent.parent
if (innereye_root / "InnerEye").is_dir():
    innereye_root_str = str(innereye_root)
    if innereye_root_str not in sys.path:
        print(f"Adding InnerEye folder to sys.path: {innereye_root_str}")
        sys.path.insert(0, innereye_root_str)
from InnerEye.Common import fixed_paths

# This must be added before all other imports because they might rely on hi-ml already, and that can optionally live
# in a submodule
fixed_paths.add_submodules_to_path()
```

**Comment by ktakeda1** [comment, marked as answer]
@furtheraway and @csmayo157... To clarify further, this type of MLOps is some of the added functionality that Azure Machine Learning provides in the cloud to help you

**Comment by csmayo157** [comment]
Thanks Anton!
Chuck
